### PR TITLE
Enum options check on StrList type parameter

### PR DESCRIPTION
**Pull request by krackerman**
fix: #696  validation on StrList with enum option in StrList class. created tests to get full coverage

### CODE REVIEW

File: src/params/str-list.js [hunk 1]
```diff
@@ -1,16 +1,25 @@
...
 
 class StrList extends List {
-  constructor(...args) {
-    super(...args);
+  constructor(name, position, opts) {
+    super(name, position, opts);
     this.items = { type: 'string' };
+
+    if (opts !== undefined && opts.enums) {
+      this.enums = new Set(opts.enums);
```
Comment (simlu) [inline src/params/str-list.js]: lets put an assert above that `Array.isArray(opts.enums)`

File: src/params/str-list.js [hunk 2]
```diff
@@ -1,16 +1,25 @@
...
+    }
   }
 
   validate(value) {
     let valid = super.validate(value);
     if (valid && (this.stringInput ? JSON.parse(value) : value).some(e => typeof e !== 'string')) {
       valid = false;
     }
+
+    if (this.enums !== undefined && !(this.stringInput ? JSON.parse(value) : value).every(val => this.enums.has(val))) {
```
Comment (simlu) [inline src/params/str-list.js]: wrap both clauses in `if (valid) {` and separate `this.stringInput ? JSON.parse(value) : value` into variable

File: test/params/str-list.spec.js [hunk 3]
```diff
@@ -4,6 +4,8 @@ const api = require('../../src/index').Api();
 describe('Testing StrList Parameter', () => {
   const queryParam = api.StrList('list', 'query');
   const jsonParam = api.StrList('list', 'json');
+  const enumBody = api.StrList('list', 'json', { enums: ['enumOne', 'enumTwo', 'enumThree'] });
```
Comment (simlu) [inline test/params/str-list.spec.js]:
```suggestion
  const queryParamEnum = api.StrList('list', 'json', { enums: ['enumOne', 'enumTwo', 'enumThree'] });
```

File: test/params/str-list.spec.js [hunk 4]
```diff
@@ -4,6 +4,8 @@ const api = require('../../src/index').Api();
...
+  const enumParam = api.StrList('list', 'query', { enums: ['enumOne', 'enumTwo', 'enumThree'] });
```
Comment (simlu) [inline test/params/str-list.spec.js]:
```suggestion
  const queryParamEnum = api.StrList('list', 'query', { enums: ['enumOne', 'enumTwo', 'enumThree'] });
```

File: test/params/str-list.spec.js [hunk 5]
```diff
@@ -36,4 +38,20 @@ describe('Testing StrList Parameter', () => {
...
+
+  it('Testing enums json parameter', () => {
+    expect(enumBody.get({
+      body: {
+        list: ['enumOne', 'enumTwo']
+      }
+    })).to.deep.equal(['enumOne', 'enumTwo']);
+  });
+
+  it('Testing valid enum query parameter', () => {
```
Comment (simlu) [inline test/params/str-list.spec.js]: ?

File: test/params/str-list.spec.js [hunk 6]
```diff
@@ -36,4 +38,20 @@ describe('Testing StrList Parameter', () => {
...
+    })).to.deep.equal(['enumOne', 'enumTwo']);
+  });
+
+  it('Testing invalid enum query params', () => {
+    expect(() => queryParamEnum.get({
+      queryStringParameters: {
+        list: '["enumOne", "enumTwo", "enumFour"]'
+      }
+    })).to.throw('Invalid Value for query-Parameter "list" provided.');
+  });
```
Comment (simlu) [inline test/params/str-list.spec.js]: there should be four tests here

File: src/params/str-list.js [hunk 7]
```diff
@@ -1,16 +1,28 @@
+const assert = require('assert');
 const List = require('./list');
 
 class StrList extends List {
-  constructor(...args) {
-    super(...args);
+  constructor(name, position, opts) {
+    super(name, position, opts);
     this.items = { type: 'string' };
+    if (opts !== undefined && opts.enums) {
```
Comment (simlu) [inline src/params/str-list.js]:
```suggestion
    if (opts !== undefined && opts.enums !== undefined) {
```

File: src/params/str-list.js [hunk 8]
```diff
@@ -1,16 +1,28 @@
...
+    if (valid) {
+      const strList = this.stringInput ? JSON.parse(value) : value;
+      if (strList.some(e => typeof e !== 'string')) {
+        valid = false;
+      }
+      if (this.enums !== undefined && !strList.every(val => this.enums.has(val))) {
+        valid = false;
+      }
     }
+
```
Comment (MacMcIrish) [inline src/params/str-list.js]: ?

File: src/params/str-list.js [hunk 9]
```diff
@@ -1,16 +1,28 @@
...
+      this.enums = new Set(opts.enums);
+    }
   }
 
   validate(value) {
     let valid = super.validate(value);
-    if (valid && (this.stringInput ? JSON.parse(value) : value).some(e => typeof e !== 'string')) {
-      valid = false;
+    if (valid) {
+      const strList = this.stringInput ? JSON.parse(value) : value;
```
Comment (MacMcIrish) [inline src/params/str-list.js]: Have a look at `geo-shape.js`, and parse value similar to there.

The checks below are easier to understand if formatted similar to `geo-shape.js` as well, try not to nest if statements if not necessary

File: src/params/str-list.js [hunk 10]
```diff
@@ -1,14 +1,26 @@
...
 const List = require('./list');
 
 class StrList extends List {
-  constructor(...args) {
-    super(...args);
+  constructor(name, position, opts) {
+    super(name, position, opts);
     this.items = { type: 'string' };
+    if (opts !== undefined && opts.enums !== undefined) {
+      assert(Array.isArray(opts.enums));
```
Comment (MacMcIrish) [inline src/params/str-list.js]:
```suggestion
      assert(Array.isArray(opts.enums));
      assert(opts.enums.every(e => typeof e === 'string));
```

File: src/params/str-list.js [hunk 11]
```diff
@@ -1,14 +1,26 @@
+const assert = require('assert');
 const List = require('./list');
 
 class StrList extends List {
-  constructor(...args) {
-    super(...args);
+  constructor(name, position, opts) {
```
Comment (MacMcIrish) [inline src/params/str-list.js]: This removes that `opts !== undefined` below
```suggestion
  constructor(name, position, opts = {}) {
```

File: test/params/str-list.spec.js [hunk 12]
```diff
@@ -4,6 +4,8 @@ const api = require('../../src/index').Api();
 describe('Testing StrList Parameter', () => {
   const queryParam = api.StrList('list', 'query');
   const jsonParam = api.StrList('list', 'json');
+  const bodyParamEnum = api.StrList('list', 'json', { enums: ['enumOne', 'enumTwo', 'enumThree'] });
+  const queryParamEnum = api.StrList('list', 'query', { enums: ['enumOne', 'enumTwo', 'enumThree'] });
```
Comment (MacMcIrish) [inline test/params/str-list.spec.js]: Change order for consistency

File: test/params/str-list.spec.js [hunk 13]
```diff
@@ -36,4 +38,36 @@ describe('Testing StrList Parameter', () => {
...
     })).to.throw('Invalid Value for json-Parameter "list" provided.');
   });
+
+  it('Testing valid enums json parameter', () => {
+    expect(bodyParamEnum.get({
+      body: {
+        list: ['enumOne', 'enumTwo']
+      }
+    })).to.deep.equal(['enumOne', 'enumTwo']);
+  });
```
Comment (MacMcIrish) [inline test/params/str-list.spec.js]: Change test order for consistency

File: src/params/str-list.js (hunk 11 above)
Comment (krackerman) [reply src/params/str-list.js]: I like this change

**Pull request merged**